File: 4 semestre/Paradigmas de Linguagem de Programação/10_outubro/aula 23-10-25/Java vs C++ vs C# vs GO vs Ruby/sistema_biblioteca.rb
=begin
  Paradigmas de Linguagem de Programação
  Aula: 23-10-25
  Slide 9: Diferença entre Java vs C++ vs C# vs GO vs Ruby
  Exemplo em Ruby - Sistema de Biblioteca
=end


class Livro
  attr_reader :titulo, :autor
  attr_accessor :emprestado

  def initialize(titulo, autor)
    @titulo = titulo
    @autor = autor
    @emprestado = false
  end

  def emprestar
    @emprestado = true
  end

  def devolver
    @emprestado = false
  end

  def to_s
    status = @emprestado ? "Emprestado" : "Disponível"
    "#{@titulo} - #{@autor} (#{status})"
  end
end

class Biblioteca
  def initialize
    @livros = []
  end

  def adicionar_livro(titulo, autor)
    @livros << Livro.new(titulo, autor)
  end

  def listar_livros
    if @livros.empty?
      puts "Nenhum livro cadastrado."
    else
      @livros.each_with_index do |livro, i|
        puts "#{i + 1}. #{livro}"
      end
    end
  end

  def buscar_por_titulo(titulo)
    livro = @livros.find { |l| l.titulo.downcase == titulo.downcase }
    if livro
      puts "Encontrado: #{livro}"
    else
      puts "Livro não encontrado."
    end
  end

  def emprestar_livro(indice)
    if indice >= 0 && indice < @livros.size
      livro = @livros[indice]
      if !livro.emprestado
        livro.emprestar
        puts "Livro emprestado com sucesso."
      else
        puts "Livro já está emprestado."
      end
    else
      puts "Índice inválido."
    end
  end

  def devolver_livro(indice)
    if indice >= 0 && indice < @livros.size
      livro = @livros[indice]
      if livro.emprestado
        livro.devolver
        puts "Livro devolvido com sucesso."
      else
        puts "Livro não está emprestado."
      end
    else
      puts "Índice inválido."
    end
  end
end

biblioteca = Biblioteca.new
opcao = nil

loop do
  puts "\n--- Menu ---"
  puts "1. Adicionar livro"
  puts "2. Listar livros"
  puts "3. Buscar por título"
  puts "4. Emprestar livro"
  puts "5. Devolver livro"
  puts "0. Sair"
  print "Escolha uma opção: "
  opcao = gets.to_i

  case opcao
  when 1
    print "Título: "
    titulo = gets.chomp
    print "Autor: "
    autor = gets.chomp
    biblioteca.adicionar_livro(titulo, autor)
  when 2
    biblioteca.listar_livros
  when 3
    print "Digite o título: "
    busca = gets.chomp
    biblioteca.buscar_por_titulo(busca)
  when 4
    biblioteca.listar_livros
    print "Digite o número do livro para emprestar: "
    indice = gets.to_i - 1
    biblioteca.emprestar_livro(indice)
  when 5
    biblioteca.listar_livros
    print "Digite o número do livro para devolver: "
    indice = gets.to_i - 1
    biblioteca.devolver_livro(indice)
  when 0
    puts "Encerrando..."
    break
  else
    puts "Opção inválida."
  end
end

File: 4 semestre/Paradigmas de Linguagem de Programação/10_outubro/aula 23-10-25/Java vs C++ vs C# vs GO vs Ruby/SistemaBiblioteca.cs
/*
Paradigmas de Linguagem de Programação
Aula: 23-10-25
Slide 9: Diferença entre Java vs C++ vs C# vs GO vs Ruby
Exemplo em C++ - Sistema de Biblioteca
*/

using System;
using System.Collections.Generic;

class Livro
{
    public string Titulo { get; private set; }
    public string Autor { get; private set; }
    public bool Emprestado { get; private set; }

    public Livro(string titulo, string autor)
    {
        Titulo = titulo;
        Autor = autor;
        Emprestado = false;
    }

    public void Emprestar()
    {
        Emprestado = true;
    }

    public void Devolver()
    {
        Emprestado = false;
    }

    public override string ToString()
    {
        return $"{Titulo} - {Autor}" + (Emprestado ? " (Emprestado)" : " (Disponível)");
    }
}

class Biblioteca
{
    private List<Livro> livros;

    public Biblioteca()
    {
        livros = new List<Livro>();
    }

    public void AdicionarLivro(string titulo, string autor)
    {
        livros.Add(new Livro(titulo, autor));
    }

    public void ListarLivros()
    {
        if (livros.Count == 0)
        {
            Console.WriteLine("Nenhum livro cadastrado.");
            return;
        }

        for (int i = 0; i < livros.Count; i++)
        {
            Console.WriteLine($"{i + 1}. {livros[i]}");
        }
    }

    public void BuscarPorTitulo(string titulo)
    {
        foreach (var livro in livros)
        {
            if (livro.Titulo.Equals(titulo, StringComparison.OrdinalIgnoreCase))
            {
                Console.WriteLine("Encontrado: " + livro);
                return;
            }
        }
        Console.WriteLine("Livro não encontrado.");
    }

    public void EmprestarLivro(int indice)
    {
        if (indice >= 0 && indice < livros.Count)
        {
            var livro = livros[indice];
            if (!livro.Emprestado)
            {
                livro.Emprestar();
                Console.WriteLine("Livro emprestado com sucesso.");
            }
            else
            {
                Console.WriteLine("Livro já está emprestado.");
            }
        }
        else
        {
            Console.WriteLine("Índice inválido.");
        }
    }

    public void DevolverLivro(int indice)
    {
        if (indice >= 0 && indice < livros.Count)
        {
            var livro = livros[indice];
            if (livro.Emprestado)
            {
                livro.Devolver();
                Console.WriteLine("Livro devolvido com sucesso.");
            }
            else
            {
                Console.WriteLine("Livro não está emprestado.");
            }
        }
        else
        {
            Console.WriteLine("Índice inválido.");
        }
    }
}

class SistemaBiblioteca
{
    static void Main(string[] args)
    {
        Biblioteca biblioteca = new Biblioteca();
        int opcao;

        do
        {
            Console.WriteLine("\n--- Menu ---");
            Console.WriteLine("1. Adicionar livro");
            Console.WriteLine("2. Listar livros");
            Console.WriteLine("3. Buscar por título");
            Console.WriteLine("4. Emprestar livro");
            Console.WriteLine("5. Devolver livro");
            Console.WriteLine("0. Sair");
            Console.Write("Escolha uma opção: ");

            if (!int.TryParse(Console.ReadLine(), out opcao))
            {
                Console.WriteLine("Entrada inválida.");
                continue;
            }

            switch (opcao)
            {
                case 1:
                    Console.Write("Título: ");
                    string titulo = Console.ReadLine();
                    Console.Write("Autor: ");
                    string autor = Console.ReadLine();
                    biblioteca.AdicionarLivro(titulo, autor);
                    break;

                case 2:
                    biblioteca.ListarLivros();
                    break;

                case 3:
                    Console.Write("Digite o título: ");
                    string busca = Console.ReadLine();
                    biblioteca.BuscarPorTitulo(busca);
                    break;

                case 4:
                    biblioteca.ListarLivros();
                    Console.Write("Digite o número do livro para emprestar: ");
